### PR TITLE
feat: add remediation report generator for PR body output

### PR DESCRIPTION
**Pull request by ruromero**
## Summary

- Add `src/remediation_report.js` with `generateReport()` and `generateDeduplicationKey()` exports
- Support per-dependency, bundled (grouped by severity), dry-run (tabular), and JSON output formats
- Include provider source attribution, CVE IDs, severity, and advisory links in all report formats
- Add deterministic deduplication key generator for PR branch/title matching

## Test plan

- [x] Per-dependency report format with multiple CVEs
- [x] Bundled report format groups by severity
- [x] Dry-run output shows tabular summary
- [x] JSON format returns valid JSON
- [x] Provider source attributed in every entry
- [x] Deduplication key generation is deterministic and stable
- [x] Edge cases (null, empty, missing groupId, advisories without URLs)
- [x] All 22 tests passing
- [x] ESLint: 0 errors

Implements [TC-5413](https://redhat.atlassian.net/browse/TC-5413)

[TC-5413]: https://redhat.atlassian.net/browse/TC-5413?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

## Summary by Sourcery

Add a remediation report generator for security dependency updates with support for multiple output formats and a deterministic deduplication key.

New Features:
- Introduce remediation_report module that generates per-dependency and bundled markdown reports for security remediations.
- Support dry-run tabular summaries and JSON output of remediation data for CLI and automation usage.
- Add a deterministic deduplication key generator for remediation entries based on dependency coordinates.

Tests:
- Add comprehensive tests covering all report formats, edge cases, and deduplication key behavior.

### CODE REVIEW

File: src/remediation_report.js [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+		if (rem.cves.length > 0) {
+			lines.push('### Vulnerabilities resolved')
+			lines.push('')
+			lines.push('| CVE | Severity | Advisory |')
+			lines.push('| --- | --- | --- |')
+			for (const cve of rem.cves) {
+				const advisoryLinks = formatAdvisoryLinks(rem.advisories)
+				lines.push(`| ${cve} | ${rem.severity} | ${advisoryLinks} |`)
+			}
+		}
```
[the remark this answers is not in the document]
Comment (ruromero) [reply src/remediation_report.js]: [sdlc-workflow/verify-pr] Classified as **suggestion** — this proposes hoisting a loop-invariant computation for performance. CONVENTIONS.md has no rules about loop-invariant extraction, and ESLint config contains no performance-related rules. Valid code quality improvement but not convention-backed. No sub-task created.

File: test/remediation_report.test.js [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+			expect(parsed).to.deep.equal(remediations)
+		})
+
+		/** Verifies that empty remediations produce empty JSON array. */
+		test('returns empty JSON array for no remediations', () => {
+			const report = generateReport([], { format: 'json' })
+
+			expect(report).to.equal('[]')
+		})
+	})
```
[the remark this answers is not in the document]
Comment (ruromero) [reply test/remediation_report.test.js:294]: [sdlc-workflow/verify-pr] Classified as **suggestion** — this proposes adding option-precedence tests. CONVENTIONS.md requires 82% line coverage (met at 98.93%) but has no rules about combinatorial option testing or precedence coverage. Valid test coverage improvement but not convention-backed. No sub-task created.

**Comment by ruromero** [comment]
[sdlc-workflow/verify-pr] Re: @sourcery-ai review — Classified as **suggestion** — the review body summarizes the same 2 issues raised as inline comments (loop-invariant hoisting and option-precedence tests). Both are valid code quality suggestions but neither matches a CONVENTIONS.md rule or established codebase pattern. No sub-task created.

**Comment by ruromero** [comment]
## Verification Report for TC-5413 (commit 431fc33)

| Check | Result | Details |
|-------|--------|---------|
| Review Feedback | PASS | 2 suggestions from sourcery-ai (loop-invariant hoisting, precedence tests); no code change requests |
| Root-Cause Investigation | N/A | No sub-tasks created |
| Scope Containment | PASS | PR files exactly match task's Files to Create (2/2) |
| Diff Size | PASS | 530 insertions across 2 new files; test:code ratio 1.8:1 |
| Commit Traceability | PASS | Single commit references TC-5413 in body |
| Sensitive Patterns | PASS | No secrets or credentials in 530 added lines |
| CI Status | PASS | All 5 checks pass (lint Node 22/24, Sourcery, PR title, commit messages) |
| Acceptance Criteria | WARN | 4/5 fully met; CVSS score absent (spec inconsistency — upstream extractor doesn't provide CVSS data) |
| Test Quality | PASS | 22 tests, all documented with JSDoc, no repetitive patterns; Eval Quality: N/A |
| Test Change Classification | ADDITIVE | New test file with 342 lines of coverage |
| Verification Commands | PASS | `npm run lint`: 0 errors; `npm test`: 22/22 passing |

### Overall: WARN

The only non-PASS finding is the CVSS score column mentioned in Acceptance Criterion 1 but absent from the implementation. This is a **spec inconsistency**, not a code defect — the upstream remediation extractor (TC-5410) does not extract CVSS data, so the report module has no CVSS values to render. The task description's template included a CVSS column, but the input data shape doesn't support it. Recommend updating the Jira task's AC1 to remove the CVSS reference, or adding CVSS extraction to the upstream extractor as a follow-up.

---
*This comment was AI-generated by [sdlc-workflow/verify-pr](https://github.com/RHEcosystemAppEng/sdlc-plugins) v0.13.8.*

File: src/remediation_report.js [hunk 3]
```diff
@@ -0,0 +1,188 @@
+/**
+ * Report generator that transforms remediation extractor output into structured
+ * markdown for PR bodies, CLI dry-run output, and JSON.
+ */
+
+const SEVERITY_ORDER = ['UNKNOWN', 'LOW', 'MEDIUM', 'HIGH', 'CRITICAL']
```
Comment (a-oren) [inline src/remediation_report.js]: src/remediation_report.js:6 defines SEVERITY_ORDER identically to src/remediation.js. This is a DRY violation, the constant should be imported from remediation.js (or extracted to a shared module) rather than duplicated.

File: src/remediation_report.js [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+			: rem.artifactId
+
+		const lines = [
+			`## Security Update: ${depName} ${rem.currentVersion} → ${rem.fixedInVersion}`,
+			'',
+			`**Provider:** ${rem.provider} | **Source:** ${rem.source}`,
+			'',
+		]
+
+		if (rem.cves.length > 0) {
```
Comment (a-oren) [inline src/remediation_report.js]: will throw a TypeError if cves is undefined or null. Same issue in generateBundledReport (line 102) with rem.cves.join(', ')


```suggestion
		if (rem.cves && rem.cves.length > 0) {
```

**Review by a-oren — CHANGES_REQUESTED** [review]
src/index.js re-exports extractRemediations from remediation.js as part of the public API surface. The new generateReport and generateDeduplicationKey are not re-exported. Every other public function in this library is accessible through the index, this omission breaks that convention and forces consumers to import from a deep path.

File: src/remediation_report.js (hunk 3 above)
Comment (ruromero) [reply src/remediation_report.js]: [sdlc-workflow/verify-pr] Classified as **code change request** — SEVERITY_ORDER is defined identically in both src/remediation.js:335 and src/remediation_report.js:6. Sub-task [TC-5513](https://redhat.atlassian.net/browse/TC-5513) created to address this feedback.

File: src/remediation_report.js (hunk 4 above)
Comment (ruromero) [reply src/remediation_report.js]: [sdlc-workflow/verify-pr] Classified as **code change request** — rem.cves access without null guard will throw TypeError if cves is undefined/null. Same issue in generateBundledReport at line 102. Sub-task [TC-5514](https://redhat.atlassian.net/browse/TC-5514) created to address this feedback.

**Comment by ruromero** [comment]
[sdlc-workflow/verify-pr] Re: @a-oren review — Classified as **code change request** — generateReport and generateDeduplicationKey are not re-exported from src/index.js, breaking the barrel-export convention established by all other public functions. Sub-task [TC-5515](https://redhat.atlassian.net/browse/TC-5515) created to address this feedback.

[TC-5515]: https://redhat.atlassian.net/browse/TC-5515?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

**Comment by ruromero** [comment]
## Verification Report for TC-5413 (commit 4bdfeba)

| Check | Result | Details |
|-------|--------|---------|
| Review Feedback | WARN | 3 code change requests from a-oren; sub-tasks [TC-5513](https://redhat.atlassian.net/browse/TC-5513), [TC-5514](https://redhat.atlassian.net/browse/TC-5514), [TC-5515](https://redhat.atlassian.net/browse/TC-5515) created |
| Root-Cause Investigation | DONE | 2 root-cause tasks: [TC-5516](https://redhat.atlassian.net/browse/TC-5516) (implement-task skill gap), [TC-5517](https://redhat.atlassian.net/browse/TC-5517) (convention gap) |
| Scope Containment | PASS | PR files match task spec (2/2 created) |
| Diff Size | PASS | 550 lines across 2 new files; test:code ratio 1.9:1 |
| Commit Traceability | PASS | Both commits reference TC-5413 |
| Sensitive Patterns | PASS | No secrets detected in 550 added lines |
| CI Status | PASS | All 5 checks pass |
| Acceptance Criteria | WARN | 4/5 met; CVSS score absent (spec inconsistency — upstream extractor doesn't provide CVSS data) |
| Test Quality | WARN | Repetitive Test Detection: WARN (dedup key + edge case tests parameterizable); Test Documentation: PASS; Eval Quality: N/A |
| Test Change Classification | ADDITIVE | New test file with 362 lines |
| Verification Commands | PASS | Lint 0 errors, 24/24 tests passing |

### Overall: WARN

**Review feedback (3 code change requests from a-oren):**
1. **TC-5513** — SEVERITY_ORDER constant duplicated between src/remediation.js and src/remediation_report.js (DRY violation)
2. **TC-5514** — `rem.cves.length` and `rem.cves.join()` will throw TypeError if cves is null/undefined
3. **TC-5515** — `generateReport` and `generateDeduplicationKey` not re-exported from src/index.js, breaking barrel-export convention

**Acceptance Criteria:** CVSS score column mentioned in AC1 but absent — upstream extractor (TC-5410) doesn't extract CVSS data. Spec inconsistency, not code defect.

**Root-cause investigation:** Two systemic improvements identified:
- **TC-5516** (implement-task skill gap): skill should search for existing symbols before re-declaring constants, and add null guards for external data property access
- **TC-5517** (convention gap): barrel-export convention is implicit but undocumented in CONVENTIONS.md

---
*This comment was AI-generated by [sdlc-workflow/verify-pr](https://github.com/RHEcosystemAppEng/sdlc-plugins) v0.13.8.*

[TC-5513]: https://redhat.atlassian.net/browse/TC-5513?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[TC-5514]: https://redhat.atlassian.net/browse/TC-5514?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[TC-5515]: https://redhat.atlassian.net/browse/TC-5515?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[TC-5516]: https://redhat.atlassian.net/browse/TC-5516?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[TC-5517]: https://redhat.atlassian.net/browse/TC-5517?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

**Comment by ruromero** [comment]
## Verification Report for TC-5413 (commit f26910c)

| Check | Result | Details |
|-------|--------|---------|
| Review Feedback | PASS | All 3 prior code change requests addressed; no new unclassified feedback |
| Root-Cause Investigation | N/A | No new sub-tasks; prior investigation completed (TC-5516, TC-5517) |
| Scope Containment | WARN | 4 files vs 2 specified; extra 2 (src/index.js, src/remediation.js) justified by review feedback sub-tasks |
| Diff Size | PASS | 584 additions across 4 files; 99.7% in task-specified files |
| Commit Traceability | PASS | All 5 commits reference TC-5413 or its sub-tasks (TC-5513, TC-5514, TC-5515) |
| Sensitive Patterns | PASS | No secrets detected in 584 added lines |
| CI Status | WARN | 3/5 checks pass, 2 pending (Node 22/24 lint+test matrix) |
| Acceptance Criteria | WARN | 4/5 met; CVSS score absent (spec inconsistency — upstream extractor doesn't provide CVSS data) |
| Test Quality | WARN | Repetitive Test Detection: WARN (3 parameterizable groups); Test Documentation: PASS (28/28 documented); Eval Quality: N/A |
| Test Change Classification | ADDITIVE | New test file with 394 lines, 28 test cases |
| Verification Commands | PASS | Lint 0 errors, 28/28 tests passing |

### Overall: WARN

**Review feedback resolved:** All 3 code change requests from a-oren (SEVERITY_ORDER DRY violation, null guard for rem.cves, missing barrel exports) have been addressed in fix commits b943125, f26910c, and 6e72189 respectively.

**Remaining WARNs:**
- **Scope Containment:** 2 extra files (src/index.js, src/remediation.js) are minimal, justified modifications from review feedback sub-tasks
- **CI Status:** Node 22/24 matrix jobs still pending at time of verification; prior run passed all 5 checks
- **Acceptance Criteria:** CVSS score in AC1 is a known spec inconsistency (upstream extractor TC-5410 doesn't extract CVSS); not a code defect

---
*This comment was AI-generated by [sdlc-workflow/verify-pr](https://github.com/RHEcosystemAppEng/sdlc-plugins) v0.13.8.*

**Review by a-oren — APPROVED** [review]
